src/control/family.lean

```diff
@@ -0,0 +1,666 @@
+/-
+Copyright (c) 2019 Simon Hudon. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Author: Simon Hudon
+-/
+import category_theory.category
+import category_theory.types
+import logic.relation
+
+/-!
+Indexed type families and their categorical structure.
+
+Features:
+
+`fam I`      : family of types and its instance of `category`
+
+Also, support functions for operating with n-tuples of types, such as:
+
+`append1 α β`    : append type J-indexed `β` family to I-indexed `α` family to obtain a `(I⊕J)`-indexed family
+`drop α`         : drop the right component of a `(I⊕J)`-indexed family
+`last α`         : take the right component of a `(I⊕J)`-indexed family
+`append_fun f g` : appends two families of functions `f` and `g`
+`drop_fun f`     : drops the right function family from a `(I⊕J)`-indexed family
+`last_fun f`     : returns the right function family of a `(I⊕J)`-indexed family
+
+Since e.g. `append1 α.drop α.last` is propositionally equal to `α` but not definitionally equal
+to it, we need support functions and lemmas to mediate between constructions.
+-/
+
+universes u v w
+
+open category_theory
+
+/-- Type family indexed by `I`; we call `fam (I ⊕ J)` the product
+of `fam I` and `fam J` (despite being formulated using a sum type) -/
+@[reducible]
+def fam (I : Type u) := I → Type u
+
+instance {I} : has_one (fam I) :=
+⟨ λ _, punit ⟩
+
+namespace fam
+
+variables {I J : Type u}
+
+/-- retain the left side of a product family -/
+def drop : fam (I ⊕ J) → fam I :=
+λ x i, x (sum.inl i)
+
+/-- retain the right side of a product family -/
+def last : fam (I ⊕ J) → fam J :=
+λ x i, x (sum.inr i)
+
+/-- combine two families with different indices their product -/
+def append1 (f : fam I) (g : fam J) : fam (I ⊕ J)
+| (sum.inl i) := f i
+| (sum.inr i) := g i
+
+end fam
+
+section pis
+
+variables {C : Type u} [𝒞 : category.{v} C]
+include 𝒞
+
+instance pi.category {α : Type w} : category $ α → C :=
+{ hom := λ X Y, Π ⦃i⦄, X i ⟶ Y i,
+  id := λ X i, 𝟙 (X i),
+  comp := λ X Y Z f g i, @f i ≫ @g i }
```

Comment:
I'm concerned here about using an implicit argument for `i` in `hom`. I tried generalizing this definition to dependent families, and realised I couldn't even express the `simp` lemmas `id_apply` and `comp_apply`, because I wanted to write things like `(𝟙 X) i = 𝟙 (X i)`.

Even ignoring any such generalization, this instance probably deserves some `simp` lemmas, and should be in `category_theory/pi/basic.lean` or somesuch.

Reply:
I pulled this instance out as #3735.